### ISSUE TITLE
We could really use a viable alternative to the slow Proto format

### ISSUE THREAD

**Issue by atyshka** [waymo-research/waymo-open-dataset#257]
I'm really happy with all the work Waymo put into developing a quality dataset and a great library to work with it, but it's extraordinarily frustrating to deal with the data format of the protos. I'm currently a student and I'd love to work with this dataset but I don't have fancy computing resources and tons of storage. I've been using GCSFuse to mount the dataset, and then I've been using an eager function within the graph to do the proto deserialization. The thing is, my model is almost entirely input bound, it's more than half idle as it waits for the poor cpu to pump out dataset frames. I am aware of the version that exists in tfds, but it has some differences that are a dealbreaker such as being in unstructured pointcloud format vs raw range-image format. The only alternative, as previously mentioned, is to do conversion offline, but again I don't have the resources to store terabytes of data. Given the Waymo challenges are now here again, it's even more important that Waymo recognize the need for a suitable, performant alternative. It wouldn't take much to convert to a tfrecord format and put it in a 3rd GCS bucket alongside the existing tar and untar'd buckets and doing so would greatly help competition participants and the research community.

**Comment by tianweiy** [comment]
Providing the files using the format defined in the challenge will be super helpful 

https://github.com/waymo-research/waymo-open-dataset/tree/master/waymo_open_dataset/latency#data-formats

**Comment by anantsubramanian** [comment]
> It wouldn't take much to convert to a tfrecord format and put it in a 3rd GCS bucket alongside the existing tar and untar'd buckets and doing so would greatly help competition participants and the research community.

Since the tfrecord format is also backed by protobufs, I wonder if it will be any faster to convert it to a tfrecord format that has similar structure? Is it possible for you to process data frames in parallel using multiple workers to feed to your model code?

> Providing the files using the format defined in the challenge will be super helpful
> 
> https://github.com/waymo-research/waymo-open-dataset/tree/master/waymo_open_dataset/latency#data-formats

I think hosting the entire dataset in that format might not be ideal, although it is nice to have the simple format as input to the model. I'm not sure, but it might be a more verbose/unstructured format for users to download and parse in their code?

It might make sense to have a couple of examples in that format to test the model for the challenge though.

**Comment by atyshka** [comment]
> Since the tfrecord format is also backed by protobufs, I wonder if it will be any faster to convert it to a tfrecord format that has similar structure

I believe so. There will be performance gains because tf.examples can work in the tensorflow graph. The problem with current protos is not they they are protos, but that they require eager execution to work. There is an overhead associated with jumping out of the computational graph to execute a python function which returns numpy arrays, and then to convert those numpy arrays into tensors in the graph.

**Comment by peisun1115** [comment]
One problem is that the tf example format is different for different models:
1. single frame model.
2. multi-frame model.
3. model with camera image.

Even if the inputs are the same, say single frame model, you might want to compute different training targets. for example, per point object type, anchor matching etc etc.

It is impossible to provide a tf-example that works for everybody. We actually do have tf example data released contributed by [lingvo](https://github.com/tensorflow/lingvo) authors.  You can use that if that fits your needs. but it is also quite slow to train directly from that as it contains all the data. It is slow to read while training.

**Comment by sabeekpradhan** [comment]
@tianweiy 

> Providing the files using the format defined in the challenge will be super helpful
> 
> https://github.com/waymo-research/waymo-open-dataset/tree/master/waymo_open_dataset/latency#data-formats

You can extract a Frame proto into the same set of key-value pairs that the latency challenge uses with the `convert_frame_to_dict` function in [utils/frame_utils.py](https://github.com/waymo-research/waymo-open-dataset/blob/ae21353bf721bf36e197654e67d482b5619a2302/waymo_open_dataset/utils/frame_utils.py#L215). I'll add a note about this in the latency README.

**Comment by tianweiy** [comment]
Thanks, that will be super helpful for us to ensure that we use the same lidar data as the evalution server

**Comment by YoushaaMurhij** [comment]
@sabeekpradhan , @tianweiy I have used `convert_frame_to_dict` function in [utils/frame_utils.py](https://github.com/waymo-research/waymo-open-dataset/blob/ae21353bf721bf36e197654e67d482b5619a2302/waymo_open_dataset/utils/frame_utils.py#L215).  to get a data_dict example and stored it as a pickle.
I am quite not sure that this will work for Latency test [here](https://github.com/waymo-research/waymo-open-dataset/blob/a19ba044e0ad42c49e95f09298b491db5d8ed05c/waymo_open_dataset/latency/wod_latency_evaluator.py#L106)
I not confused right now about the correct format! Can you please provide a correct example that can be worked as `eval_data`?

**Comment by sabeekpradhan** [comment]
Hi Youshaa. The main difference from what you seem to be doing is that we don't save the dictionary returned by `convert_frame_to_dict` using pickle. Instead, we save the dictionary into separate numpy arrays using [`np.save`](https://numpy.org/doc/stable/reference/generated/numpy.save.html) (where each key-value pair will have the value saved in a `.npy` file whose basename is `<key>.npy`, e.g. `TOP_RANGE_IMAGE_FIRST_RETURN.npy`). The directory structure we use on the eval server is save the `.npy` files for a frame at `<root_dir>/<frame.context.name>/<frame.timestamp_micros>`, and each such directory will have one `.npy` file for each entry in the dict returned by `convert_frame_to_dict` (plus the versions with `_1` and `_2` appended to the key for the values from the previous two frames). But if you're running this locally, note that you can pass any input directory you want to [`process_example`](https://github.com/waymo-research/waymo-open-dataset/blob/a19ba044e0ad42c49e95f09298b491db5d8ed05c/waymo_open_dataset/latency/wod_latency_evaluator.py#L42) so long as it contains `.npy` files for all the data fields your module uses.

**Comment by YoushaaMurhij** [comment]
Thanks @sabeekpradhan for your clear answer! I manged to solve it following your instructions!

**Comment by atyshka** [comment]
@peisun1115 Have you looked at all into tf.io.decode_proto? I'm currently trying to get that working, as I think that could be a viable solution to retaining the current proto format while being able to use graph execution. My current stumbling block is the function requires proto description files, which are not currently generated during compilation. I'm trying to see if I can generate those files from the wheels by reconstructing them from the dataset_pb2.py file or add their generation to the compilation process.
